### PR TITLE
RavenDB-17719 MergedDocumentReplicationCommand dispose replication item only on committed transaction

### PR DESCRIPTION
**Pull request by haludi**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17719

### Additional description

In the MergedDocumentReplicationCommand.ExecuteCmd we put each replicated item in `using` so at the end of the execution, all the items were disposed.
When a merged transaction including MergedDocumentReplicationCommand fails on another command
the next attempt in RunEachOperationIndependently of MergedDocumentReplicationCommand will use disposed items.

The dispose of ReplicationItem do something only for:
`ReplicationBatchItem.ReplicationItemType.Attachment`
`ReplicationBatchItem.ReplicationItemType.AttachmentTombstone`
`ReplicationBatchItem.ReplicationItemType.RevisionTombstone`

In addition, as I see for now there is only one allocation on the same ByteStringContext and it is for a size bigger than `ByteStringContext.MinBlockSizeInBytes`.
The memory that was released is used for small strings and it is stored for later reuse (_internalReusableStringPool).
This is why we don't see this cause any issue.

### Type of change
- Bug fix

### How risky is the change?
- Moderate 

### Backward compatibility
- Not relevant

### Is it platform specific issue?
- No

### Documentation update
- No documentation update is needed 

### Testing 
- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?
- No

### UI work
- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 1]
```diff
@@ -1275,7 +1275,21 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
                     var lastTransactionMarker = 0;
                     var handledAttachmentStreams = new HashSet<Slice>(SliceComparer.Instance);
                     context.LastDatabaseChangeVector = context.LastDatabaseChangeVector ?? DocumentsStorage.GetDatabaseChangeVector(context);
-                    foreach (var item in _replicationInfo.ReplicatedItems)
+
+                    var tx = context.Transaction.InnerTransaction.LowLevelTransaction;
+                    var replicationItems = _replicationInfo.ReplicatedItems;
+                    tx.OnDispose += _ =>
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1281]: What if this fail for the second time as well? would we leak here?

Comment (haludi) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1281]: In this case, the `TransactionOperationsMerger` will set the `MergedDocumentReplicationCommand` task as failed and the task that waiting for it (`ReceiveSingleDocumentsBatch`) will throw and finally will release the whole context.
This is the same as it handles it now for untreated items.

Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1281]: okay, make sense. Nice catch :)

**Review by ayende — COMMENTED** [review]
This is something that we have to look at for other commands as well.
The killer for us is that this may also cause us to access memory that we freed, no?
That can lead to AccessViolationExceptions that are _very_ hard to track down.

**Comment by ayende** [comment]
And agree with @karmeli87 - @haludi this is a *great* catch

**Pull request merged**